server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java

```diff
@@ -169,48 +174,52 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             if (shardRouting.unassignedInfo() != null) {
                 unassignedInfoToXContent(shardRouting.unassignedInfo(), builder);
             }
-            if (currentNode != null) {
+
+            if (this.currentNode != null){
                 builder.startObject("current_node");
-                {
-                    discoveryNodeToXContent(currentNode, true, builder);
-                    if (shardAllocationDecision.getMoveDecision().isDecisionTaken()
-                        && shardAllocationDecision.getMoveDecision().getCurrentNodeRanking() > 0) {
-                        builder.field("weight_ranking", shardAllocationDecision.getMoveDecision().getCurrentNodeRanking());
-                    }
+                discoveryNodeToXContent(this.currentNode, true, builder);
+                if (shardAllocationDecision.getMoveDecision().isDecisionTaken()
+                    && shardAllocationDecision.getMoveDecision().getCurrentNodeRanking() > 0) {
+                    builder.field("weight_ranking", shardAllocationDecision.getMoveDecision().getCurrentNodeRanking());
                 }
                 builder.endObject();
             }
-            if (this.clusterInfo != null) {
-                builder.startObject("cluster_info");
-                {
-                    // This field might be huge, TODO add chunking support here
-                    ChunkedToXContent.wrapAsToXContent(clusterInfo).toXContent(builder, params);
-                }
-                builder.endObject(); // end "cluster_info"
-            }
-            if (shardAllocationDecision.isDecisionTaken()) {
-                shardAllocationDecision.toXContent(builder, params);
+
+            return builder;
+        }),
+            this.clusterInfo != null
+                ? Iterators.concat(
+                    Iterators.single((builder, p) -> builder.startObject("cluster_info")),
+                    this.clusterInfo.toXContentChunked(params),
+                    Iterators.<ToXContent>single((builder, p) -> builder.endObject())
```

Comment:
There's some helpers for this case:

```suggestion
                    ChunkedToXContentHelper.startObject("cluster_info"),
                    this.clusterInfo.toXContentChunked(params),
                    ChunkedToXContentHelper.endObject()
```